docs/history-service.md

```diff
@@ -0,0 +1,274 @@
+- Timer fired
+
+In general, the History Service implements these state transitions in the same way, in many cases using the [same code path](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/api/update_workflow_util.go#L37):
+
+```mermaid
+sequenceDiagram
+participant shard as History Shard
+participant persistence as Persistence
+note over shard: RPC received or timer fired
+shard->>shard: 
+note over shard: In-memory:<br>create new History Event(s)<br>update Mutable State<br>create History Task(s)
+shard->>persistence: append new History Event(s)
+shard->>persistence: (transactionally) update <br>Mutable State and add History Task(s)
+```
+
+<details>
+<summary><i>Code entrypoints</i></summary>
+
+These state transitions are performed using the generic utility function [GetAndUpdateWorkflowWithNew](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/api/update_workflow_util.go#L37). The utility function receives a function that performs the updates to mutable state specific to the particular state transition, and then performs the two commits to persistence via
+[`UpdateWorkflowExecutionAsActive`](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/workflow/context.go#L459), ultimately calling the persistence backends, e.g.
+[sql](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/common/persistence/sql/execution.go#L339) and [cassandra](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/common/persistence/cassandra/execution_store.go#L142).
+
+- When handling a `ScheduleActivityTask` command from a Worker, a task is [added to in-memory mutable state](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/workflow/task_generator.go#L419) representing the Transfer Task that must be persisted.
+
+Call sites:
+
+- Handling an RPC from the User Application, for example [`SignalWorkflow`](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/api/signalworkflow/api.go#L54).
+
+- Handling a `WorkflowTaskCompleted` RPC from the Worker in [`handleWorkflowTaskCompleted`](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/workflow_task_handler_callbacks.go#L690)
+
+- Timer task processor: [`executeUserTimerTimeoutTask`](https://github.com/temporalio/temporal/blob/ef49189005b5323c532264287af6c08a447aab8a/service/history/timer_queue_active_task_executor.go#L136) handles a `workflow.sleep()` timer set in user code by appending `TimerFired`, and `WorkflowTaskScheduled` events to Workflow History, and creating a Transfer Task for the Workflow Task.
+
+</details>
+<br>
+<br>
+
+#### Consistency guarantees
+
+- Consistency of Mutable State and History Tasks is achieved via use of database transactions.
```

Comment:
It worth to explain/point out (may be not here) why we store 6 different objects in `executions` table.